jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -182,41 +186,44 @@ public static URL computePURootURL(URL pxmlURL, String descriptorLocation) throw
+        } else if("zip".equals(protocol) ||
+                  "jar".equals(protocol) ||
+                  "wsjar".equals(protocol)) {
+            // e.g. file:/foo/bar.jar!/META-INF/persistence.xml
+            // "zip:" URLs require additional handling.
+            String spec = "zip".equals(protocol)
+                ? "file:" + pxmlURL.getFile()
+                : pxmlURL.getFile();
+
             int separator = spec.lastIndexOf("!/");
-            if (separator == -1) {
-                separator = spec.length() - 1;
+
+            // It could be possible for a "zip:" or "wsjar:" URL to not have
+            // an entry! In that case we take the root of the archive.
+            String file = separator == -1 ? spec : spec.substring(0, separator);
+            String entry = separator == -1 ? "" : spec.substring(separator + 2);
+
+            // The jar file or directory whose META-INF directory contains
+            // the persistence.xml file is termed the root of the persistence
+            // unit. (JPA Spec, 8.2)
+            if (!entry.endsWith(descriptorLocation)) {
+                // Shouldn't happen unless we have a particularly tricky
+                // classloader - which we're not obligated to support.
+                return null;
             }
-            result = new File(spec.substring(0, separator++)).toURL();
-        } else if("wsjar".equals(protocol)) { // NOI18N
-            // e.g. wsjar:file:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            // but WS gives use jar:file:..., so we need to match it.
-            String spec = pxmlURL.getFile();
-            int separator = spec.lastIndexOf("!/");
-            if (separator == -1) {
-                separator = spec.length();
-            } else {
-                // If this doesn't reference a war file with a properly located persistence.xml,
-                // then chop off everything after the "!/" marker and assume it is a normal jar.
-                // Else, if the wsjar URL references a file with a ".war" extension, and its entry
-                // starts with WEB-INF/classes/, then the calculated persistence unit root should
-                // be wsjar:path/to/a.war!/WEB-INF/classes/ as per JPA 2.1 Spec section 8.2 "Persistence Unit Packaging".
-                separator += 2;
-                // Filter out invalid scenarios such as wsjar:file:/a/path/to/my.war!/foo/WEB-INF/classes/META-INF/persistence.xml
-                if (spec.regionMatches(true, separator - 6, ".war", 0, 4) &&
-                        spec.regionMatches(true, separator, WEBINF_CLASSES_STR, 0, WEBINF_CLASSES_LEN)) {
-                    separator += WEBINF_CLASSES_LEN;
-                }
+
+            String rootEntry = entry.substring(0, entry.length() - descriptorLocation.length());
+
+            // Since EclipseLink is a reference implementation, let's validate
+            // the produced root!
+            if (!isValidRootInArchive(file, rootEntry)) {
```

Comment:
I was thinking about this a bit more and I think throwing an exception should be better. Take a look at

https://github.com/eclipse-ee4j/eclipselink/blob/ac5913f745ebbe74d1dcd3e38ffea1baba208141/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/config/persistenceunit/PersistenceUnitImpl.java#L53-L60

What if we changed that to this instead:

``` java
        URL puURL = PersistenceUnitImpl.class.getClassLoader().getResource(persistenceFactoryResource);
        try{
            puURL = PersistenceUnitProcessor.computePURootURL(puURL, persistenceFactoryResource);
        } catch (Exception ex){
            throw PersistenceUnitLoadingException.exceptionProcessingPersistenceUnit(puURL, ex);
        }
```

The other usage already takes care of exceptions so that's covered:
https://github.com/eclipse-ee4j/eclipselink/blob/ac5913f745ebbe74d1dcd3e38ffea1baba208141/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java#L332

Then, we just need an exception to throw from ```computePURootURL()```. I think adding a new exception message here is warranted. I couldnt find an existing one that makes any sense. Take a look at:
https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/exceptions/ValidationException.java 
https://github.com/eclipse-ee4j/eclipselink/blob/master/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/exceptions/i18n/ValidationExceptionResource.java
to see how we can add a new exception. I'm thinking something like ```{ "7357", "The following URL is not valid {0}"}``` maybe? Then we can throw that exception from ```computePURootURL()```? Thoughts? :)

Reply:
> Thoughts? :)

Sure, I'll get on it tomorrow.

Reply:
Done.